summoner-cli/src/Summoner/Template/GitHub.hs

```diff
@@ -121,49 +121,121 @@ gitHubFiles Settings{..} = concat
         , ""
-        , "    - uses: actions/setup-haskell@v1"
+        , "    - uses: actions/setup-haskell" <> ghcActionsSetupHaskellVersion
+        , "      id: setup-haskell-cabal"
         , "      name: Setup Haskell"
         , "      with:"
         , "        ghc-version: ${{ matrix.ghc }}"
         , "        cabal-version: ${{ matrix.cabal }}"
         , ""
-        , "    - uses: actions/cache@v1"
+        , "    - name: Freeze"
+        , "      run: |"
+        , "        cabal v2-freeze"
+        , ""
+        , "    - uses: actions/cache" <> ghcActionsCacheVersion
         , "      name: Cache ~/.cabal/store"
         , "      with:"
-        , "        path: ~/.cabal/store"
-        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-cabal"
+        , "        path: ${{ steps.setup-haskell-cabal.outputs.cabal-store }}"
+        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-${{ hashFiles('cabal.project.freeze') }}"
         , ""
         , "    - name: Build"
         , "      run: |"
-        , "        " <> cabalUpdate
+        , "        " <> cabalConfigure
         , "        " <> cabalBuild
         , ""
         , "    - name: Test"
         , "      run: |"
         , "        " <> cabalTest
+        , ""
         ]
 
+    ghActionsStack :: [Text]
+    ghActionsStack =
+        [ "  stack:"
+        , "    name: stack / ghc ${{ matrix.ghc }}"
+        , "    runs-on: ubuntu-latest"
+        , "    strategy:"
+        , "      matrix:"
+        , "        stack: [" <> quote defaultStack <> "]"
+        , "        ghc:"
+        ]
+        <> map (indent 10 <>) ghActionsVersions
+        <>
+        [ ""
+        , "    steps:"
+        , "    - uses: actions/checkout" <> ghcActionsCheckoutVersion
+        , "      if: github.event.action == 'opened' || github.event.action == 'synchronize' || github.event.ref == 'refs/heads/master'"
+        , ""
+        , "    - uses: actions/setup-haskell" <> ghcActionsSetupHaskellVersion
+        , "      name: Setup Haskell Stack"
+        , "      with:"
+        , "        ghc-version: ${{ matrix.ghc }}"
+        , "        stack-version: ${{ matrix.stack }}"
+        , ""
+        , "    - uses: actions/cache" <> ghcActionsCacheVersion
+        , "      name: Cache ~/.stack"
+        , "      with:"
+        , "        path: ~/.stack"
+        , "        key: ${{ runner.os }}-${{ matrix.ghc }}-stack"
+        , ""
+        , "    - name: Build"
+        , "      run: |"
+        , "        " <> stackBuild
+        , ""
+        , "    - name: Test"
+        , "      run: |"
+        , "        " <> stackTest
+        ]
 
     ghActionsVersions :: [Text]
     ghActionsVersions = map
         (\ghc -> "- " <> quote (showGhcVer ghc))
         settingsTestedVersions
 
+    ghActionsExcludes :: [Text]
+    ghActionsExcludes =
+        let latestVersion = viaNonEmpty last $ sort settingsTestedVersions
```

Comment:
Here you could sort in descending order. You can use `sortWith Down` for that 🙂 
In this case, you would be able to use `head` instead, and it will be faster.

I wonder if we can use `defaultGhc` here, @chshersh ? It is always the `maxBound`, and currently, it is not possible to create the project without using this GHC 🤔

Reply:
I personally like the approach of sorting in descending order. It's short, more efficient that sort plus filter and does not depend on the assumption that `defaultGhc` is always used and the latest in case that changes in the future.

Reply:
@vrom911 @gdziadkiewicz Let's proceed with sorting in the `descending` order ⏬ Maybe we won't always use `defaultGhc` 🙂  But it's something to keep in mind for future

Reply:
It turned out that without `filter` I don't need the latest version, just the versions to exclude, so I went with `drop 1` after `sortWith Down`.